search/search_quality/merge_samples.py

```diff
@@ -0,0 +1,181 @@
+    """
+    for field in ["query", "locale"]:
+        if a[field] != b[field]:
+            return False
+
+    pos_key = "position"
+    bad_point = {"x": -1, "y": -1}
+    a_pos = a[pos_key] or bad_point
+    b_pos = b[pos_key] or bad_point
+    if not points_equal(a_pos, b_pos):
+        return False
+
+    vp_key = "viewport"
+    if not rects_equal(a[vp_key], b[vp_key]):
+        return False
+
+    return True
+
+
+def results_equal(a, b):
+    """
+    Returns whether two found results are equal before assessing.
+    Does not take relevance into account.
+    """
+    pos_key = "position"
+    name_key = "name"
+    hn_key = "houseNumber"
+    types_key = "types"
+    if a[name_key] != b[name_key]:
+        return False
+    if not points_equal(a[pos_key], b[pos_key]):
+        return False
+    if a[hn_key] != b[hn_key]:
+        return False
+    if sorted(a[types_key]) != sorted(b[types_key]):
+        return False
+    return True
+
+
+def greedily_match_results(a, b):
+    match_to_a = [-1] * len(a)
+    match_to_b = [-1] * len(b)
+    for i in range(len(a)):
+        if match_to_a[i] >= 0:
+            continue
+        for j in range(len(b)):
+            if match_to_b[j] >= 0:
+                continue
+            if results_equal(a[i], b[j]):
+                match_to_a[i] = j
+                match_to_b[j] = i
+                break
+    return match_to_a, match_to_b
+
+
+def merge_relevancies(a, b):
+    RELEVANCIES = ["harmful", "irrelevant", "relevant", "vital"]
+    id_a = RELEVANCIES.index(a)
+    id_b = RELEVANCIES.index(b)
+    if id_a > id_b:
+        a, b = b, a
+        id_a, id_b = id_b, id_a
+    if id_a == 0 and id_b <= 1:
+        return a, True
+    if id_a == 1 and id_b <= 2:
+        return b, True
+    if id_a > 1:
+        return b, True
+    return a, False
+
+
+def merge_two_samples(a, b, line_number):
+    if not samples_equal(a, b):
+        raise Exception(f"Tried to merge two non-equivalent samples, line {line_number}")
+
+    useless_key = "useless"
+    useless_a = useless_key in a and a[useless_key]
+    useless_b = useless_key in b and b[useless_key]
+
+    if useless_a != useless_b:
+        print(line_number, "useless:", useless_a, useless_b)
+
+    if useless_a and not useless_b:
+        return b
+    if useless_b:
+        return a
+
+    # Both are not useless.
+
+    res_key = "results"
+    match_to_a, match_to_b = greedily_match_results(a[res_key], b[res_key])
+
+    lst = [x for x in match_to_a if x < 0]
+
+    rel_key = "relevancy"
+
+    c = copy.deepcopy(a)
+    c[res_key] = []
+
+    for i, j in enumerate(match_to_a):
+        j = match_to_a[i]
+        if j < 0:
+            continue
+        res = a[res_key][i]
+        ra = a[res_key][i][rel_key]
+        rb = b[res_key][j][rel_key]
+        rc, ok = merge_relevancies(ra, rb)
+        res[rel_key] = rc
+        if not ok:
+            print(line_number, ra, rb, a[res_key][i]["name"])
+            pass
```

Comment:
no need in pass

Reply:
That was a reminder. Turned out we're ok with a simple print here.
done